packages/casting/src/follower-cosmjs.js

```diff
@@ -196,85 +224,89 @@ export const makeCosmjsFollower = (
-    assert(
-      result.length >= dataPrefixBytes.length,
-      X`result too short for data prefix ${dataPrefixBytes}`,
-    );
-    assert(
-      arrayEqual(result.subarray(0, dataPrefixBytes.length), dataPrefixBytes),
-      X`${result} doesn't start with data prefix ${dataPrefixBytes}`,
-    );
-    return result.slice(dataPrefixBytes.length);
+    return { value, height: result.height };
   };
 
   /**
-   * @param {number} [height]
-   * @returns {Promise<Uint8Array>}
+   * @param {number} height
+   * @returns {Promise<QueryAbciResponse>}
    */
   const getProvenDataAtHeight = async height => {
-    return retryGetDataAndStripPrefix(
-      async (endpoint, storeName, storeSubkey) => {
-        const queryClient = await provideQueryClient(endpoint);
-        return E(queryClient).queryVerified(storeName, storeSubkey, height);
-      },
-    );
+    return retryGetPrefixedData(async (endpoint, storeName, storeSubkey) => {
+      const queryClient = await provideQueryClient(endpoint);
+      const value = await E(queryClient).queryVerified(
+        storeName,
+        storeSubkey,
+        height,
+      );
+      return { value, height };
+    });
   };
 
   /**
-   * @param {number} height
+   * @param {number} [height] desired height, or the latest height if not set
+   * @returns {Promise<QueryAbciResponse>}
    */
   const getUnprovenDataAtHeight = async height => {
-    return retryGetDataAndStripPrefix(
-      async (endpoint, storeName, storeSubkey) => {
-        const client = await provideTendermintClient(endpoint);
-        const response = await client.abciQuery({
-          path: `store/${storeName}/key`,
-          data: storeSubkey,
-          height,
-          prove: false,
-        });
-        if (response.code !== 0) {
-          throw new Error(`Tendermint ABCI query failed: ${response.log}`);
-        }
-        const { value } = response;
-        return value;
-      },
-    );
+    return retryGetPrefixedData(async (endpoint, storeName, storeSubkey) => {
+      const client = await provideTendermintClient(endpoint);
+      const response = await client.abciQuery({
+        path: `store/${storeName}/key`,
+        data: storeSubkey,
+        height,
+        prove: false,
+      });
+      if (response.code !== 0) {
+        throw new Error(`Tendermint ABCI query failed: ${response.log}`);
+      }
+      if (!response.height) {
+        throw new Error('No query height returned');
+      }
+      return {
+        value: response.value,
+        height: response.height,
+      };
+    });
   };
 
   /**
-   * @param {number} blockHeight
+   * @param {number} [blockHeight] desired height, or the latest height if not set
+   * @returns {Promise<QueryAbciResponse>}
    */
   const tryGetDataAtHeight = async blockHeight => {
     if (proof === 'strict') {
+      // we have to know this in order to construct a valid return
+      if (blockHeight === undefined) {
+        blockHeight = await getBlockHeight();
```

Comment:
for the `strict` case, this is repeating the `getBlockHeight()` more than before. I figure this all is good factoring though, because there should be a way to get the data _and height_ even for proven data.

I don't know what the performance impact of this is in practice but there are other ways to work around the Stargate deficiency with less perf regression. 

1. have `if (proof === 'strict')` to get the block height at higher levels in the call stack
2. memoize `getBlockHeight()` with an expiration of expected block time (6s)

Reply:
That PR lacks the `queryVerified` case but https://github.com/cosmos/cosmjs/pull/1278#issuecomment-1326527195 indicates that was just step 1. I've linked to your PR as a place to find whether it's resolved.